src/aabb.rs

```diff
@@ -0,0 +1,175 @@
+}
+
+/// A trait implemented by things which can be bounded by an `AABB`.
+pub trait Bounded {
+    fn aabb(&self) -> AABB;
+}
+
+impl AABB {
+    /// Creates a new `AABB` with the given bounds.
+    pub fn new(min: Point3<f32>, max: Point3<f32>) -> AABB {
+        AABB {
+            min: min,
+            max: max,
+        }
+    }
+
+    /// Creates a new empty `AABB`.
+    pub fn empty() -> AABB {
+        AABB {
+            min: Point3::new(f32::INFINITY, f32::INFINITY, f32::INFINITY),
+            max: Point3::new(f32::NEG_INFINITY, f32::NEG_INFINITY, f32::NEG_INFINITY),
+        }
+    }
+
+    /// Returns true if the `Point3` is inside the `AABB`.
+    pub fn contains(&self, p: &Point3<f32>) -> bool {
+        p.x >= self.min.x && p.x <= self.max.x && p.y >= self.min.y && p.y <= self.max.y &&
+        p.z >= self.min.z && p.z <= self.max.z
+    }
+
+    /// Returns true if the `Point3` is approximately inside the `AABB`
+    /// with respect to some `epsilon`.
+    pub fn approx_contains_eps(&self, p: &Point3<f32>, epsilon: f32) -> bool {
+        (p.x - self.min.x) > -epsilon && (p.x - self.max.x) < epsilon &&
+        (p.y - self.min.y) > -epsilon && (p.y - self.max.y) < epsilon &&
+        (p.z - self.min.z) > -epsilon && (p.z - self.max.z) < epsilon
+    }
+
+    /// Returns a new minimal `AABB` which contains both this `AABB` and `other`.
+    pub fn union_aabb(&self, other: &AABB) -> AABB {
+        AABB::new(Point3::new(self.min.x.min(other.min.x),
+                              self.min.y.min(other.min.y),
+                              self.min.z.min(other.min.z)),
+                  Point3::new(self.max.x.max(other.max.x),
+                              self.max.y.max(other.max.y),
+                              self.max.z.max(other.max.z)))
+    }
+
+    /// Returns a new minimal `AABB` which contains both this `AABB` and `other`.
+    pub fn union_point(&self, other: &Point3<f32>) -> AABB {
+        AABB::new(Point3::new(self.min.x.min(other.x),
+                              self.min.y.min(other.y),
+                              self.min.z.min(other.z)),
+                  Point3::new(self.max.x.max(other.x),
+                              self.max.y.max(other.y),
+                              self.max.z.max(other.z)))
+    }
+
+    /// Returns a new minimal `AABB` which contains both this `AABB` and `other`.
+    pub fn union_bounded<T: Bounded>(&self, other: &T) -> AABB {
+        self.union_aabb(&other.aabb())
+    }
+
+    /// Returns the size of this `AABB` in all three dimensions.
+    pub fn size(&self) -> Vector3<f32> {
+        self.max - self.min
+    }
+
+    /// Returns the center point of the `AABB`.
+    pub fn center(&self) -> Point3<f32> {
+        self.min + (self.size() / 2.0)
+    }
+
+    /// Returns the total surface area of this `AABB`.
+    pub fn surface_area(&self) -> f32 {
+        let size = self.size();
+        2.0 * size.x * size.y + size.x * size.z + size.y * size.z
```

Comment:
Shouldn't it be `2.0 * size.x * size.y + 2.0 * size.x * size.z + 2.0 * size.y * size.z`?